File: lesson12/ScalableImageViewTest/app/src/main/java/com/chinalwb/scalableimageviewtest/view/ScalableImageView.java
package com.chinalwb.scalableimageviewtest.view;

import android.animation.ObjectAnimator;
import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.support.annotation.Nullable;
import android.support.v4.view.GestureDetectorCompat;
import android.util.AttributeSet;
import android.util.Log;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.widget.OverScroller;

import com.chinalwb.scalableimageviewtest.R;

public class ScalableImageView extends View implements
            GestureDetector.OnGestureListener, GestureDetector.OnDoubleTapListener, Runnable {
    private Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);

    private int sw;
    private int width, height, bitmapWidth, bitmapHeight;
    private float offsetX, offsetY;

    private Bitmap bitmap;
    private boolean isBig;
    private float fraction;
    private ObjectAnimator animator;
    private int smallScale = 1, bigScale = 10;
    private float translateY, oy;

    private GestureDetectorCompat gestureDetectorCompat;
    private OverScroller overScroller;


    float dy;

    public ScalableImageView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        gestureDetectorCompat = new GestureDetectorCompat(getContext(), this);
        gestureDetectorCompat.setOnDoubleTapListener(this);
        overScroller = new OverScroller(getContext());
        sw = Resources.getSystem().getDisplayMetrics().widthPixels;
        bitmap = Utils.getBitmap(getResources(), R.drawable.timg, sw / bigScale);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        this.width = w;
        this.height = h;
        this.bitmapWidth = this.bitmap.getWidth();
        this.bitmapHeight = this.bitmap.getHeight();
        offsetX = (this.width - this.bitmapWidth) / 2f;
        offsetY = (this.height - this.bitmapHeight) / 2f;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        dy = -offsetY * fraction;
        canvas.translate(0, dy + translateY);
        Log.e("XX", "translation y == " + (dy + translateY));

        float scale = smallScale + (bigScale - smallScale) * fraction;
        canvas.scale(scale, scale, offsetX + bitmapWidth / 2f, offsetY);
        canvas.drawBitmap(bitmap, offsetX, offsetY, paint);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        return gestureDetectorCompat.onTouchEvent(event);
    }

    public float getFraction() {
        return fraction;
    }

    public void setFraction(float fraction) {
        this.fraction = fraction;
        invalidate();
    }

    public ObjectAnimator getAnimator() {
        if (animator == null) {
            animator = ObjectAnimator.ofFloat(this, "fraction", 0, 1);
        }
        return animator;
    }

    @Override
    public boolean onDown(MotionEvent e) {
        return true;
    }

    @Override
    public void onShowPress(MotionEvent e) {

    }

    @Override
    public boolean onSingleTapUp(MotionEvent e) {
        return false;
    }

    @Override
    public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
        if (!isBig) {
            return false;
        }

        translateY -= distanceY;
        translateY = Math.min(translateY, 0);
        translateY = Math.max(translateY, height - bitmapHeight * bigScale);
        invalidate();
        return false;
    }

    @Override
    public void onLongPress(MotionEvent e) {

    }

    @TargetApi(16)
    @Override
    public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
        if (!isBig) {
            return false;
        }
        Log.e("XX", "-------------- onfling --------- " + (dy + translateY));
        overScroller.fling(0, (int) (dy + translateY), (int) velocityX, (int) velocityY,
                0, 0,
                height - bitmapHeight * bigScale,
                0
        );

        oy = dy + translateY;
        postOnAnimation(this);
        return false;
    }

    @Override
    public boolean onSingleTapConfirmed(MotionEvent e) {
        return false;
    }

    @Override
    public boolean onDoubleTap(MotionEvent e) {
        if (this.isBig) {
            translateY = 0;
            getAnimator().reverse();
        } else {
            getAnimator().start();
        }
        this.isBig = !this.isBig;
        return false;
    }

    @Override
    public boolean onDoubleTapEvent(MotionEvent e) {
        return false;
    }

    @TargetApi(16)
    @Override
    public void run() {
        if (overScroller.computeScrollOffset()) {
            int cy = overScroller.getCurrY();
//            Log.e("XX", "cy == " + cy + ", oy == " + oy);
            float dy = oy - cy;
//            Log.e("XX", "dy == " + dy);
            oy = cy;
            translateY -= dy;
//            Log.e("XX", "translate y == " + translateY);

            translateY = Math.min(translateY, 0);
            translateY = Math.max(translateY, height - bitmapHeight * bigScale);
            invalidate();
            postOnAnimation(this);
        }
    }
}

File: lesson12/ScalableImageViewTest/app/src/main/java/com/chinalwb/scalableimageviewtest/view2/ScalableImageView.kt
package com.chinalwb.scalableimageviewtest.view2

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ObjectAnimator
import android.content.Context
import android.graphics.*
import android.os.Build
import android.support.annotation.RequiresApi
import android.support.v4.view.GestureDetectorCompat
import android.util.AttributeSet
import android.util.Log
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.ScaleGestureDetector
import android.view.View
import android.widget.OverScroller
import com.chinalwb.scalableimageviewtest.R
import com.chinalwb.scalableimageviewtest.view.Utils
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

class ScalableImageView(context: Context, attributeSet: AttributeSet) : View(context,
        attributeSet), Runnable {

    private var paint: Paint = Paint(Paint.ANTI_ALIAS_FLAG)
    var bitmap: Bitmap? = null
    var bWidth = 0
    var bHeight = 0
    var bLeft = 0F
    var bTop = 0F
    var smallScale = 1F
    var bigScale = 1F
    private val overScale = 2F
    private var cx = 0F
    var cy = 0F
    var isBig = false
    private var currentScale = 0F
        set(value) {
            field = value
            invalidate()
        }

    var objectAnimator: ObjectAnimator? = null

    private var gestureListener = MyGestureListener()
    var gestureDetector = GestureDetectorCompat(context, gestureListener)

    var offsetX = 0F
    var offsetY = 0F

    var overScroller = OverScroller(context)

    private var scaleGestureListener = MyScaleGestureListener()
    var scaleGestureDetector = ScaleGestureDetector(context, scaleGestureListener)

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        bitmap = loadResource(R.drawable.s, Utils.dp2px(300).toInt())

        bWidth = bitmap!!.width
        bHeight = bitmap!!.height

        bLeft = (width - bWidth) / 2F
        bTop = (height - bHeight) / 2F

        smallScale = if (bWidth >= bHeight) width / bWidth.toFloat() else height / bHeight.toFloat()
        bigScale = if (bWidth >= bHeight) height / bHeight.toFloat() else width / bWidth.toFloat()
        bigScale *= overScale
        currentScale = smallScale

        cx = width / 2F
        cy = height / 2F
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)

        canvas!!.drawColor(Color.LTGRAY)

        var scaleFraction = (currentScale - smallScale) / (bigScale - smallScale)
        canvas.translate(offsetX * scaleFraction, offsetY * scaleFraction)
        canvas.scale(currentScale, currentScale, cx, cy)
        canvas.drawBitmap(bitmap!!, bLeft, this.bTop, paint)
    }

    private fun loadResource(resId: Int, targetWidth: Int): Bitmap {
        var options = BitmapFactory.Options()
        options.inJustDecodeBounds = true
        BitmapFactory.decodeResource(resources, resId, options)
        options.inJustDecodeBounds = false
        options.inDensity = options.outWidth
        options.inTargetDensity = targetWidth

        return BitmapFactory.decodeResource(resources, resId, options)
    }


    override fun onTouchEvent(event: MotionEvent?): Boolean {
        var result = scaleGestureDetector.onTouchEvent(event)
        if (!scaleGestureDetector.isInProgress) {
            result = gestureDetector.onTouchEvent(event)
        }
        return result
    }


    @RequiresApi(Build.VERSION_CODES.JELLY_BEAN)
    override fun run() {
        if (overScroller.computeScrollOffset()) {
            offsetX = overScroller.currX.toFloat()
            offsetY = overScroller.currY.toFloat()
            invalidate()
            postOnAnimation(this)
        }
    }


    /**
     * Double Click listener
     */
    inner class MyGestureListener : GestureDetector.OnGestureListener, GestureDetector
    .OnDoubleTapListener {
        override fun onShowPress(e: MotionEvent?) {
        }

        override fun onSingleTapUp(e: MotionEvent?): Boolean {
            return false
        }

        override fun onDown(e: MotionEvent?): Boolean {
            return true
        }

        @RequiresApi(Build.VERSION_CODES.JELLY_BEAN)
        override fun onFling(down: MotionEvent?, e2: MotionEvent?, velocityX: Float, velocityY: Float):
                Boolean {
            if (!isBig) return false

            overScroller.fling(
                    offsetX.toInt(),
                    offsetY.toInt(),
                    velocityX.toInt(),
                    velocityY.toInt(),
                    -((bWidth * bigScale) - width).toInt() / 2,
                    ((bWidth * bigScale) - width).toInt() / 2,
                    -((bHeight * bigScale) - height).toInt() / 2,
                    ((bHeight * bigScale) - height).toInt() / 2
            )

            postOnAnimation(this@ScalableImageView)
            return false
        }

        override fun onScroll(down: MotionEvent?, event: MotionEvent?, distanceX: Float, distanceY:
        Float): Boolean {
            if (!isBig) return false

            offsetX -= distanceX
            offsetY -= distanceY
            fixOffsets()
            invalidate()
            return false
        }

        override fun onLongPress(e: MotionEvent?) {
        }


        // --- OnDoubleTapListener callbacks

        override fun onDoubleTap(e: MotionEvent?): Boolean {
            isBig = !isBig

            Log.w("XX", "OnDoubleTap")
            if (isBig) {
                // Small to Big
                offsetX = (e!!.x - cx) - (e.x - cx) / smallScale * bigScale
                offsetY = (e.y - cy) - (e.y - cy) / smallScale * bigScale
                fixOffsets()
                getAnimator().start()
            } else {
                // Big to small
                var objectAnimator = getAnimator()
                objectAnimator.reverse()
            }
            return false
        }

        private fun getAnimator(): ObjectAnimator {
            if (objectAnimator == null) {
                objectAnimator = ObjectAnimator.ofFloat(this@ScalableImageView, "currentScale", 0F)
            }

            objectAnimator!!.setFloatValues(smallScale, bigScale)
            return objectAnimator!!
        }

        override fun onDoubleTapEvent(e: MotionEvent?): Boolean {
            return false
        }

        override fun onSingleTapConfirmed(e: MotionEvent?): Boolean {
            return false
        }

        private fun fixOffsets() {
            offsetX = if (offsetX < 0) {
                max(offsetX, -abs(bWidth * bigScale - width) / 2)
            } else {
                min(offsetX, abs(bWidth * bigScale - width) / 2)
            }

            offsetY = if (offsetY < 0) {
                max(offsetY, -abs(bHeight * bigScale - height) / 2)
            } else {
                min(offsetY, abs(bHeight * bigScale - height) / 2)
            }
        }
    } // #eof MyGestureListener

    /**
     * Scale listener
     */
    inner class MyScaleGestureListener : ScaleGestureDetector.OnScaleGestureListener {
        var initScale = 1F
        override fun onScaleBegin(detector: ScaleGestureDetector?): Boolean {
            initScale = currentScale
            return true
        }

        override fun onScale(detector: ScaleGestureDetector?): Boolean {
            currentScale = initScale * detector!!.scaleFactor
            isBig = currentScale > smallScale
            invalidate()
            return false
        }

        override fun onScaleEnd(detector: ScaleGestureDetector?) {

        }
    }
}